refactor(sdk): Replace `quick-lru` with `lru-cache`

mondoreale opened the pull request:
_(Hopefully one of many small PRs.)_

## Summary

`quick-lru` is an ES-only package and in order to build with it for CJS we currently move it (raw ESM) over (to `/vendor`) and transform to CommonJS as part of SDK's build process.

`lru-cache` is already being used by DHT and it organically comes in 2 flavours: CJS and ESM. No magic required, and no magic = less maintenance.

## Changes

- Replace references.
- Address differences in tests, in types, and in logic.
- Fix and annotate tests for `Mapping`.
- Remove "vendor" magic (no longer required).

## Checklist before requesting a review

- [x] ~~Is this a breaking change? If it is, be clear in summary.~~
- [x] Read through code myself one more time.
- [x] Make sure any and all `TODO` comments left behind are meant to be left in.
- [x] Has reasonable passing test coverage?
- [x] ~~Updated changelog if applicable.~~
- [x] ~~Updated documentation if applicable.~~

## Remaining todos
- [x] Fix tests

mondoreale committed:
Drop vendor and the vendor-hack

mondoreale committed:
Swap `quick-lru` with `lru-cache`

The motivation here is that unlike `quick-lru` `lru-cache` comes in 2 formats we're interested in: cjs and esm.

Plus, DHT already uses `lru-cache` so it already fits the stack.

mondoreale committed:
Fix package jsons

mondoreale committed:
Fix `MaxSizedSet`

Incorrect param!

mondoreale commented on packages/sdk/test/unit/customMatchers.test.ts:
Let me revert this one. Most likely node v18 vs v20 issue.

mondoreale replied:
Can't get this to pass locally. "Me" problem, it seems. CI sees no issues.

mondoreale commented:
Oh, this needs a bump. Lotta conflicts.

teogeb commented on packages/sdk/src/utils/Mapping.ts:
```diff
@@ -19,6 +19,14 @@ interface Item<K, V> {
     value: V
 }
 
+interface LRU<K extends string, V> {
```
Could this pick something like `type AbstractMap = Pick<Map, ...>`. In the `else` branch we create an instance, which is not a LRU (just a simple Map) and therefore could rename e.g. to `AbstractMap`.

mondoreale replied:
Types of `Map` and `LRUCache` are incompatible. I'll keep the explicit matching-all type and change its name to `CacheMap`.

teogeb commented on packages/sdk/test/unit/Mapping.test.ts:
```diff
@@ -97,25 +97,35 @@ describe('Mapping', () => {
 
     it('max size', async () => {
-        const SIZE = 3
+        const maxSize = 3
```
Good rename. Let's use naming style of constants.

```suggestion
        const MAX_SIZE = 3
```

teogeb commented on packages/sdk/src/utils/Mapping.ts:
```diff
@@ -45,9 +53,9 @@ export class Mapping<K extends LookupKeyType, V> {
-                maxSize: opts.maxSize,
-                maxAge: opts.maxAge
+            this.delegate = new LRUCache<string, Item<K, V>>({
+                max: opts.maxSize,
```
Would it make sense to use `maxSize` option of LRUCache instead of the `max` option? Both of these options seem to limit the item count, but `max` option allocates the storage pre-emptively. This may be more performant, but also bit wasteful, as very typically our caches are not in anywhere near the maximum limit.

mondoreale replied:
I guess. That's how it used to be. I'll adjust.

teogeb commented on packages/sdk/test/unit/Mapping.test.ts:
```diff
@@ -97,25 +97,35 @@ describe('Mapping', () => {
             await mapping.get(['foo', id])
         }
-        expect(valueFactory).toHaveBeenCalledTimes(5)
+        expect(valueFactory).toHaveBeenCalledTimes(maxSize + 1 + 3)
```
Why this is now 7 instead of 5? Assumed it should be the initial 3 evaluations + evaluation of `foo-1` + re-evaluation of dropped `foo0` 

Also: could use `MAX_SIZE` here in place of 3?

mondoreale replied:
This tests does the following
| | Operation | Slot 0 | Slot 1 | Slot 2 |
| - | --------- | ------ | ------ | ------ |
| 1 | `get('foo',  0)` | foo0|||
| 2 | `get('foo',  1)` | foo1|foo0|
| 3 | `get('foo',  2)` | foo2|foo1|foo0|
| 4 | `get('foo', -1)` | foo-1|foo2|foo1|
| 5 | `get('foo',  0)` | foo0|foo-1|foo2|
| 6 | `get('foo',  1)` | foo1|foo0|foo-1|
| 7 | `get('foo',  2)` | foo2|foo1|foo0|

Each row is a hit to the value factory because we happen to always hit an entry that's either not yet in the collection or has just been discarded.

teogeb replied:
Thanks, great explanation 👍

mondoreale committed:
Clarify type names

LRUCache isn't a Map so I use an abstract type for it.

mondoreale committed:
Use `maxSize` for new LRUs

This makes the code skip pre-allocation. Less performant, but more memory-friendly.

mondoreale committed:
Update tests

teogeb reviewed (approved):
Good improvement 👍

mondoreale commented:
I'll merge it in real quick. Cheers!

mondoreale commented:
For the record, size of the SDK browser build (streamr-sdk.web.min.js):
- before 5.07MB
- after 5.06MB

As expected, `quick-lru` is gone, and `lru-cache` did not add anything on top of what's already been there.

Just FYI.
